cpp/game_renderer.cpp

```diff
@@ -0,0 +1,253 @@
+	delete plaintexture_frag;
+	delete teamcolor_frag;
+	delete alphamask_vert;
+	delete alphamask_frag;
+
+	// Renderer keybinds
+	// TODO: a renderer settings struct
+	// would allow these to be put somewher better
+	auto &global_input_context = engine->get_input_manager().get_global_context();
+	global_input_context.bind(input::action_t::TOGGLE_BLENDING, [this](const input::action_arg_t &) {
+		this->settings.terrain_blending = !this->settings.terrain_blending;
+	});
+	global_input_context.bind(input::action_t::TOGGLE_UNIT_DEBUG, [this](const input::action_arg_t &) {
+		this->settings.draw_debug = !this->settings.draw_debug;
+
+		// TODO remove this hack
+		UnitAction::show_debug = !UnitAction::show_debug;
+	});
+
+	log::log(MSG(dbg) << "Loaded Renderer");
+}
+
+GameRenderer::~GameRenderer() {
+	// oh noes, release hl3 before that!
+	delete this->gaben;
+
+	delete texture_shader::program;
+	delete teamcolor_shader::program;
+	delete alphamask_shader::program;
+}
+
+bool GameRenderer::on_draw() {
+	Engine &engine = Engine::get();
+
+	// draw terrain
+	GameMain *game = engine.get_game();
+	if (game) {
+
+		// draw gaben, our great and holy protector, bringer of the half-life 3.
+		gaben->draw(coord::camgame{0, 0});
+
+		// TODO move render code out of terrain
+		if (game->terrain) {
+			game->terrain->draw(&engine, &this->settings);
+		}
+
+		if (this->settings.draw_grid) {
+			this->draw_debug_grid();
+		}
+	}
+	return true;
+}
+
+void GameRenderer::draw_debug_grid() {
+	Engine &e = Engine::get();
+
+	coord::camgame camera = coord::tile{0, 0}.to_tile3().to_phys3().to_camgame();
+
+	int cam_offset_x = util::mod(camera.x, e.engine_coord_data->tile_halfsize.x * 2);
+	int cam_offset_y = util::mod(camera.y, e.engine_coord_data->tile_halfsize.y * 2);
+
+	int line_half_width = e.engine_coord_data->window_size.x / 2;
+	int line_half_height = e.engine_coord_data->window_size.y / 2;
+
+	// rounding so we get 2:1 proportion needed for the isometric perspective
+
+	if (line_half_width > line_half_height * 2) {
+
+		if (line_half_width & 1) {
+			line_half_width += 1; // round up if it's odd
+		}
+
+		line_half_height = line_half_width / 2;
+
+	} else {
+		line_half_width = line_half_height * 2;
+	}
+
+	// quantity of lines to draw to each side from the center
+	int k = line_half_width / (e.engine_coord_data->tile_halfsize.x);
+
+	int tilesize_x = e.engine_coord_data->tile_halfsize.x * 2;
+	int common_x   = cam_offset_x + e.engine_coord_data->tile_halfsize.x;
+	int x0         = common_x     - line_half_width;
+	int x1         = common_x     + line_half_width;
+	int y0         = cam_offset_y - line_half_height;
+	int y1         = cam_offset_y + line_half_height;
+
+	glLineWidth(1);
+	glColor3f(0.0, 0.0, 0.0);
+	glBegin(GL_LINES); {
+
+		for (int i = -k; i < k; i++) {
+				glVertex3f(i * tilesize_x + x0, y1, 0);
+				glVertex3f(i * tilesize_x + x1, y0, 0);
+
+				glVertex3f(i * tilesize_x + x0, y0 - 1, 0);
+				glVertex3f(i * tilesize_x + x1, y1 - 1, 0);
+		}
```

Comment:
yes, but please blame me for not having a better renderer finished already. gnargh.